lib/swm050/timer.c

```diff
@@ -0,0 +1,457 @@
+		SYSCTL_SYS_CFG_1 |= (1 << 16);
+	} else {
+		/* Drop to the closest even value. */
+		SYSCTL_SYS_CFG_1 &= TIMER_DIV_MASK | ((~0x01 & div) << 16);
+	}
+}
+
+/**
+ * Enables or disables the timer.
+ * @param timer Select timer @ref timer_select
+ * @param en Enable or disable the timer
+ */
+void timer_enable(uint32_t timer, bool en)
+{
+	if (en) {
+		TIMER_CTRL(timer) |= TIMER_CTRL_EN;
+	} else {
+		TIMER_CTRL(timer) &= ~TIMER_CTRL_EN;
+	}
+}
+
+/**
+ * Enables or disables the timer's internal clock.
+ * @param timer Select timer @ref timer_select
+ * @param en Enable or disable the internal clock
+ */
+void timer_clock_enable(uint32_t timer, bool en)
+{
+	if (timer == TIMER_SE1) {
+		if (en) {
+			SYSCTL_SYS_CFG_1 |= SYSCTL_SYS_CFG_1_TIMERSE1;
+		} else {
+			SYSCTL_SYS_CFG_1 &= ~SYSCTL_SYS_CFG_1_TIMERSE1;
+		}
+	} else {
+		if (en) {
+			SYSCTL_SYS_CFG_1 |= SYSCTL_SYS_CFG_1_TIMERSE0;
+		} else {
+			SYSCTL_SYS_CFG_1 &= ~SYSCTL_SYS_CFG_1_TIMERSE0;
+		}
+	}
+}
+
+/**
+ * Selects the mode of operation.
+ * @param timer Select timer @ref timer_select
+ * @param mode The mode of operation @ref timer_operation_modes
+ */
+void timer_operation_mode(uint32_t timer, enum timer_operation_modes mode)
+{
+	TIMER_CTRL(timer) &= TIMER_OPERATION_MODE_MASK | (mode << 4);
+}
+
+/**
+ * Selects the output mode.
+ * Only used in counter mode.
+ * When done counting, the pin can be set to no output,
+ * to invert the current pin level, to set the pin high,
+ * or to set the pin low.
+ * @note Be sure to set the alternate functions of the timer pins
+ *	with @ref syscon_sel_af() and disable SWD on those pins
+ *	with @ref syscon_sel_swd() as needed.
+ * @param timer Select timer @ref timer_select
+ * @param mode The output mode @ref timer_output_modes
+ */
+void timer_output_mode(uint32_t timer, enum timer_output_modes mode)
+{
+	TIMER_CTRL(timer) &= TIMER_OUTPUT_MODE_MASK | (mode << 12);
+}
+
+/**
+ * Selects the initial output level.
+ * Only used in counter and PWM modes.
+ * @param timer Select timer @ref timer_select
+ * @param level The initial output level @ref timer_level
+ */
+void timer_output_level(uint32_t timer, enum timer_level level)
+{
+	TIMER_OUTPVAL(timer) = level;
+}
+
+/**
+ * Selects the edge mode.
+ * @param timer Select timer @ref timer_select
+ * @param mode The edge mode @ref timer_edge_modes
+ */
+void timer_edge_mode(uint32_t timer, enum timer_edge_modes mode)
+{
+	if (mode) {
+		TIMER_CTRL(timer) |= TIMER_CTRL_TMOD;
+	} else {
+		TIMER_CTRL(timer) &= ~TIMER_CTRL_TMOD;
+	}
+}
+
+/**
+ * Selects the loop mode.
+ * This has no use in PWM mode.
+ * In loop mode with counter mode, the counter will constantly loop.
+ * In loop mode with the capture modes, the values will be captured
+ * again and again. In single mode, these operations happen only once.
+ * @param timer Select timer @ref timer_select
+ * @param mode The loop mode @ref timer_loop_modes
+ */
+void timer_loop_mode(uint32_t timer, enum timer_loop_modes mode)
+{
+	if (mode) {
```

Comment:
I still feel we must have misunderstood each other somewhere that you are so wrapped up in using enums, and then treating them as booleans.  for enumerations of only two choices in other families, we get functions that do "set mode A" (timer_set_loop_mode) and "set mode B" (timer_disable_loop_mode) (which is verbose and wordy, sure)   All of this code will fall apart if anyone ever adds to the enum.  I know this is "complete" so I'm going to merge it as is, but it's.... weird to try it like this, and I don't really understand the motivation.  Using a switch also works, though is marginally bigger when compiled.

Reply:
What about timer_enable_loop_mode(uint32t timer, bool en)?